cpp/ql/src/Documentation/CommentedOutCode.qll

```diff
@@ -47,12 +47,27 @@ private predicate looksLikeCode(string line) {
   )
 }
 
+/**
+ * Holds if there is a preprocessor directive on the line indicated by
+ * `f` and `line`.
+ */
+private predicate preprocLine(File f, int line) {
+  exists(PreprocessorDirective pd, Location l |
+    pd.getLocation() = l and
+    l.getFile() = f and
+    l.getStartLine() = line
+  )
+}
+
 /**
  * The line of a C++-style comment within its file `f`.
  */
 private int lineInFile(CppStyleComment c, File f) {
   f = c.getFile() and
-  result = c.getLocation().getStartLine()
+  result = c.getLocation().getStartLine() and
+
+  // Ignore comments on the same line as a preprocessor directive.
+  not preprocLine(f, result)
```

Comment:
Why make this exception only for C++-style comments? This pattern is used in at least one C-style comment in our own codebase. You can find it with `git grep '#endif.*/\* #'` from Bash.

Reply:
This logic actually covers an edge case.  The main exclusion is in the constructor of `CommentBlock`, to catch comments that begin on the same line as a preprocessor directive, e.g.:
```
#endif /* #ifdef MYMACRO */
```
or indeed:
```
#endif /* #ifdef MYMACRO
more comment here
*/
```

The edge case involves CPP-style comments as the query has logic that merges consecutive CPP-style comments into one.  For example:
```
// first
// second
#endif // third
// fourth
```
is considered a single four-line comment.  The logic here eliminates the line `//third` _before_ merging, so that the above is now considered to be two comments (one consisting of the first two lines, and one of just the fourth).  There's a test case for this, admittedly narrow, behaviour.

Having said all this it looks like I forgot to add a test case for a CPP-style comment.  I'll do that now.

Reply:
Done.